### ISSUE TITLE
Memory leak/major performance issue with pose detection

### ISSUE THREAD

**Issue by jilinzheng** [AI-Coach-PT/ExerSights#70]
No memory issues when all detectPose calls are commented out, but very significant and increasing-over-time bug occurs when it is in use...
A guess is after the requestAnimationFrame call is made, perhaps it never actually gets stopped...

**Comment by jilinzheng** [comment]
Two possible causes as of time of writing:

1. improper cancellation of requestAnimationFrame as mentioned already (but only used in newer model, so could be an issue but not likely the sole cause)
2. (more likely) not actually closing pose landmarker model every time detectPose returns (if it ever returns...) and reinstantiating it over and over

**Comment by Anish701** [comment]
I cannot replicate this bug on my laptop even when I tried testing using the Memory snapshots. Thus, I removed myself as an assignee.

**Comment by jilinzheng** [comment]
Anish found singleton pattern to work on old model (Pose), but does not work for new model (TasksVision)

**Comment by jilinzheng** [comment]
Still another issue with the old model is that the webcam stays on, and runs into an issue where the page cannot fetch the video source...

**Comment by jilinzheng** [comment]
<img width="1470" alt="Screenshot 2024-11-14 at 19 44 09" src="https://github.com/user-attachments/assets/bcaf1bf5-8165-4373-a1c3-cbb6a70ed0ed">
<img width="1470" alt="Screenshot 2024-11-14 at 19 44 18" src="https://github.com/user-attachments/assets/be74af11-adc1-4bc6-b488-cf954d00b380">

**Comment by Anish701** [comment]
fixed the issue in the old Pose model in https://github.com/AI-Coach-PT/ExerSights/commit/b6bc516bbe452ccf915d2223f8ba29cbb5c2972d

I did this by only instantiating the pose model once (making it into a singleton) rather than instantiating it everytime detectPose gets called

**Comment by jilinzheng** [comment]
Kept only the component mounting detectPose call in commit [aa893a9](https://github.com/AI-Coach-PT/ExerSights/commit/aa893a9fc0fb5e065c09427062770d77e739fbc7), and memory leak seems to be solved by the singleton pattern attempted here too. Do we really need the detectPose calls and/or stopping the webcam stream in the other places (currently commented out)?

If not, it seems that the problem left is cleaning up the model (and/or segmentation masks?) if the user refreshes the page (unmounts and remounts, but no cleanup is currently performed for unmounting stage). Perhaps these API references will be of use...

- [PoseLandmarkerResult]https://ai.google.dev/edge/api/mediapipe/js/tasks-vision.poselandmarkerresult
- [MPMask.close(), MPMask is the wrapper under the hood of PoseLandmarkerResult's segmenationMasks property] https://ai.google.dev/edge/api/mediapipe/js/tasks-vision.mpmask#mpmaskclose

Two more things to note:

- PoseLandmarker itself does **not** have a close method - instead, it is the PoseLandmarkerResults that have the close method...
- Quoting from the MPMask.close() API reference: 

> Note that this method does not free masks that are owned by the C++ Task, as these are freed automatically once you leave the MediaPipe callback. Additionally, some shared state is freed only once you invoke the Task's close() method.
